display: add display_show

Finomnis opened the pull request:
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/79798.

Indicates to the display that all previous writes should be flushed to the display.

This adds support for display drivers that internally contain a double-buffering or latching mechanism.
Currently the only way to implement those displays is to present all `write`s to the user immediately, which could cause tearing artifacts for multi-write renders. (it's quite common for UI managers to render to partial framebuffers in low-memory situations. LVGL, for example, recommends 1/10th of the screen size)

Open questions:
- Name of the 'end-of-frame' display driver api function?
  - `display_show` (current PR code)
  -  `display_present` (could be confused with "display is present")
  - `display_flush` (sounds like it's optional)
  - `display_finalize_frame` (might be a little too verbose)
  - `display_flip` (probably not because could be confused with x or y pixel order setting, like, 'flipping' the image)
  - `display_pageflip` (might be too usecase specific)

Finomnis committed:
display: add `display_show`

Introduces support for double-buffered/latched displays.
Currently, every write has to be presented to the user immediately,
which negates the advantage of latched displays to prevent frame tearing.

Signed-off-by: Martin Stumpf <finomnis@gmail.com>

pdgendt commented on include/zephyr/drivers/display.h:
```diff
@@ -279,6 +279,29 @@ static inline int display_read(const struct device *dev, const uint16_t x,
+		(struct display_driver_api *)dev->api;
+
+	if (api->finalize_frame == NULL) {
+		return 0;
```
```suggestion
		return -ENOSYS;
```

Finomnis replied:
My thought was that if a driver does not require any special operation to finalize the frame, it can simply not implement this function and it will be successful.

faxe1008 replied:
+1 on the concern of trickeling LVGL functionality into the display driver api.

Finomnis replied:
Fixed

pdgendt reviewed (changes requested):
Introducing an API change should be accompanied with at least one driver that implements it.

That being said, this would tie the display driver closely to the lvgl way of working, which might require some more discussion.

Finomnis commented:
> Introducing an API change should be accompanied with at least one driver that implements it.

I don't think there is an upstream driver that requires this right now :/
An FPGA driver in our own project requires it.

> That being said, this would tie the display driver closely to the lvgl way of working, which might require some more discussion.

I kind of disagree with this (that it ties to LVGL, i'm of course open to discussions). I think every display works in frames, and every frame has an end.

An alternative API would be `display_start_frame` and `display_end_frame`, but I don't really know a usecase where `start_frame` would be necessary. A display usually knows that a frame started when the first `write` arrives.

Finomnis commented:
The problem I have is that regardless of LVGL, the `display_write` function nowhere states that it should be the only write per frame. And as it does not, there needs to be some way to signal double-buffered/latched displays that the modified content should now be displayed. This functionality is simply missing.

Finomnis commented:
> Introducing an API change should be accompanied with at least one driver that implements it.

Implemented it in `display_dummy` driver.

Finomnis commented:
Is the name bothering? Would `display_flush` be better than `display_finalize_frame`?

pdgendt reviewed (changes requested):
This PR should be split into different commits:

- Introduce API function
- Implement in dummy driver
- Calling from LVGL

pdgendt commented:
> The problem I have is that regardless of LVGL, the `display_write` function nowhere states that it should be the only write per frame. And as it does not, there needs to be some way to signal double-buffered/latched displays that the modified content should now be displayed. This functionality is simply missing.

@danieldegrasse @jfischer-no can you weigh in here?

Finomnis commented:
> Introducing an API change should be accompanied with at least one driver that implements it.

> This PR should be split into different commits

@pdgendt I feel like those are contradictory statements :) I'm happy to split it, as long as I don't have to merge them again because of the 'should be accompanied with at least one driver' requirement.

Finomnis committed:
drivers: display_dummy: implement `display_show`

Implements the `display_show` API function for the `display_dummy`
driver.

Signed-off-by: Martin Stumpf <finomnis@gmail.com>

pdgendt commented:
> > Introducing an API change should be accompanied with at least one driver that implements it.
> 
> > This PR should be split into different commits
> 
> @pdgendt I feel like those are contradictory statements :) I'm happy to split it, as long as I don't have to merge them again because of the 'should be accompanied with at least one driver' requirement.

I am not contradicting myself: 1 PR, 3 commits please

Finomnis committed:
lvgl: call `display_show` at the end of a frame

In frames with multiple writes (officially supported through
`CONFIG_LV_Z_VDB_SIZE`) the display needs to be signalled that the
current frame is over and the content should be displayed.
This allows displays to present the UI without tearing artifacts.

Signed-off-by: Martin Stumpf <finomnis@gmail.com>

Finomnis commented:
Aaaah. Now it makes sense. Thanks for clarifying.

Finomnis commented:
> Why are the commits and PR comments from different accounts/persons?

Work email / private email. Is this a problem?

JarmouniA commented:
> I once tried (in a different PR) with `Signed-off-by: Finomnis<finomnis@gmail.com>`, but it didn't work for some reason.

@Finomnis it should be `Signed-off-by: Martin Stumpf <finomnis@gmail.com>`

Finomnis commented:
I can't switch

> > I once tried (in a different PR) with `Signed-off-by: Finomnis<finomnis@gmail.com>`, but it didn't work for some reason.
> 
> @Finomnis it should be `Signed-off-by: Martin Stumpf <finomnis@gmail.com>`

So which one is the Problem? The `Signed-off-by` or the commit author? I can change the first one of course, but changing the second one to my private email at work would be annoying :/ and then you have the situation where the commit author does not match the `Signed-off-by`, which most likely will cause a CI error, right?

Finomnis commented:
> > I once tried (in a different PR) with `Signed-off-by: Finomnis<finomnis@gmail.com>`, but it didn't work for some reason.
> 
> @Finomnis it should be `Signed-off-by: Martin Stumpf <finomnis@gmail.com>`

Like this?

danieldegrasse commented:
> @danieldegrasse @jfischer-no can you weigh in here?

I'd like to better understand exactly what the case for a flush would be. Typically when using LVGL with a 1/10 buffer size, you need to write to the display graphics RAM when the buffer is presented, because the only RAM on the system actually capable of storing the full framebuffer is the graphics RAM on the display controller. In other cases (like the eLCDIF), the *driver* must define an internal framebuffer, and update it with the contents of the partial framebuffer each time LVGL sends a new one.

Typically the way display controllers deal with the possibility of tearing is via a TE (tearing effect) signal, which indicates when the display controller starts scanning a new frame from graphics ram to the LCD. Perhaps your FPGA could use something similar?

Finomnis commented:
> I'd like to better understand exactly what the case for a flush would be. Typically when using LVGL with a 1/10 buffer size, you need to write to the display graphics RAM when the buffer is presented, because the only RAM on the system actually capable of storing the full framebuffer is the graphics RAM on the display controller. In other cases (like the eLCDIF), the _driver_ must define an internal framebuffer, and update it with the contents of the partial framebuffer each time LVGL sends a new one.
> 
> Typically the way display controllers deal with the possibility of tearing is via a TE (tearing effect) signal, which indicates when the display controller starts scanning a new frame from graphics ram to the LCD. Perhaps your FPGA could use something similar?

@danieldegrasse There's actually two sources of synchronization that could cause tearing:

- VSync; synchronization from the display to the frame buffer - If the 'swapping' of the content is not synchronized to the vertical sync of the scanline (I guess this is the tearing effect signal you are talking about), a random tear line can appear anywhere in the imageTearing 
- Synchronization from the UI framework to the frame buffer - If we have multiple writes per frame, the information about which the last write is has to be propagated to the driver. Otherwise you could get tearing at the write chunk borders, if one write gets displayed during one frame and the next write (which is intended for the same frame) gets displayed at the next frame.

Even with a Tearing Effect/VSYNC signal, the chunk tearing would not be eliminated. The information about which write belongs to which frame, and at which write a new frame begins is simply missing from the display driver interface.

If LVGL renders in 1/10th screen chunks, all the driver sees is a continuous stream of `write` calls. So if it wants to use a VSYNC signal for tearing prevention, it needs to wait after the correct `write` call, which is currently not known to the driver.

To give you more detail about our concrete usecase, we have an FPGA with two frame buffers on it. The scanning speed of the FPGA is fixed (the UI is just an overlay to an FPGA-internal video stream), so we cannot sync the display to the UI, we need to sync the UI to the display. Further, our internal RAM is not enough to hold an entire frame, so we need to utilize LVGL partial renders.
We cannot utilize direct rendering, because for technical reasons this would be too slow. So we do need a partial frame buffer on the MCU.

Our procedure is:
1. Display scans from one FPGA buffer, while our UI can write to the other FPGA buffer.
2. UI keeps sending `write` calls to our driver, which sends the data to the off-screen FPGA framebuffer
3. Once the UI is done with a frame, we request an FPGA frame buffer swap and wait for it to happen (like, wait for VSYNC)
4. Repeat, now rendering to the other buffer while the FPGA scans from the new frame.

For step 3, we need two sync sources. We need to know:
- which `write` is the last of the frame
- when the `vsync` happens

The `vsync` part could be done using a tearing effect signal which you mention, but the information which write is the last of the frame is simply missing from the display driver API, so we have no way to prevent the chunk tearing.

Hopefully that kind of explains the problem.

danieldegrasse commented:
Yeah, this is a great explanation- thank you for laying it out. It seems like LVGL has `lv_display_flush_is_last` for this use case. Personally I see the reasoning for this API- I'm not sure how else we can do this effectively with partial rendering. Maybe we could add a flag to `display_write`? I think a new API is better here though, since we don't break downstream applications that way.

Finomnis commented:
> Maybe we could add a flag to `display_write`? I think a new API is better here though, since we don't break downstream applications that way.

Same thought. It is also be backwards compatible if we added it to the `desc` argument of `display_write`, but that feels like the wrong place.

faxe1008 commented:
Thanks for the explanation @Finomnis. Change itself looks good. Can you maybe add a short sentence about this to the v4.0 release notes?

danieldegrasse reviewed (commented):
Considering this more, we need to clearly define when a user needs to call the `display_flush` API. I agree that there is potentially a need for this change, but we have an issue here because it fundamentally represents a different method of managing displays than we currently have in Zephyr today, right? Currently the user can assume that calling `display_write` will result in data  being written to their display, but with this change that is no longer true.

If I understand the description of your downstream application right, no data would actually appear on the display until `display_flush` was called.

This changes `display_write` to effectively "queue" writes to the display, then `display_flush` passes those writes to the display.

One way I could see handling this would be to change the documentation around `display_write` to state that writes to the display may occur immediately or be queued, and a user must call `display_flush` to send all queued writes to the display.

I don't like this though,  because this means all downstream applications need to start calling `display_flush` to use the API consistently. In some ways I would prefer if we added a parameter to the `display_write` API call that acts as a hint to the driver that the display should be flushed. This would break downstream applications, but it would do so in a way that was easy to fix. I could see users easily missing the new requirement to call `display_flush`, since it wouldn't be *required* for most displays, just required by the API definition.

@faxe1008 or @pdgendt do you have any thoughts here? I would also add that I agree if we make this API change there needs to be a *full* implementation of it. I think this could actually be done in the eLCDIF driver (or another driver using full framebuffers) using logic like so:
- if `display_write` is called, simply queue the write into a framebuffer
- when `display_flush` is called, pass that framebuffer to the eLCDIF hardware so it renders to the screen.

To be clear, I don't have an issue with the concept of changing the display API to deal with this issue- I just want to make sure we fully understand the implications of doing so, and communicate the change to downstream users effectively.

faxe1008 commented:
@danieldegrasse I agree that it is kind of ambigous if a write takes immediate effect or not.  I like the extra API since it does not break downstream. Would it maybe be feasable to add a flag to the display capabilities `screen_info` field? The user still has to account for it but does not have to go hunting inside drivers sources to understand what the display wants.

danieldegrasse commented:
> @danieldegrasse I agree that it is kind of ambigous if a write takes immediate effect or not. I like the extra API since it does not break downstream. Would it maybe be feasable to add a flag to the display capabilities `screen_info` field? The user still has to account for it but does not have to go hunting inside drivers sources to understand what the display wants.

Yeah, I like this a lot. @Finomnis would you like some help getting this supported in a driver like the eLCDIF? We need to have a case of this being used in tree, and need to update the display sample (`samples/drivers/display`) as well to use it. I think adding a flag in `screen_info` makes sense, then the display driver sample can have additional handling with `display_flush` if that flag is set (same thing for LVGL).

Finomnis commented:
I kind of thought the "double buffer" capability is exactly that, but it turns out to be something weird and obscure.

There's also a use of lv_display_flush_is_last in another eink display case already where blanking gets used in an (imo) incorrect way. This could also be removed after this change and instead be implemented through a flush.

Should we rework those? What's the point of the existing double buffer capability?

faxe1008 commented:
Docs state that "The display has two alternating RAM buffers", which imo is not necessarily the same thing. I could not find any intree usage of this flag. Adding a flag with an unambigous explanation would be good I think.

jfischer-no commented:
> I kind of thought the "double buffer" capability is exactly that, but it turns out to be something weird and obscure.

IIRC some EPD use double buffer to determine what area is to update.

> There's also a use of lv_display_flush_is_last in another eink display case already where blanking gets used in an (imo) incorrect way. This could also be removed after this change and instead be implemented through a flush.

Yes, that can be used with some EPD displays that supports partial writes to display controller memory. Not sure why LVGL API is mixed here, this is just some graphics library.

> 
> Should we rework those? What's the point of the existing double buffer capability?

SCREEN_INFO_DOUBLE_BUFFER? I do not see it used in the tree.

Finomnis commented:
> Yeah, I like this a lot. @Finomnis would you like some help getting this supported in a driver like the eLCDIF? We need to have a case of this being used in tree, and need to update the display sample (`samples/drivers/display`) as well to use it. I think adding a flag in `screen_info` makes sense, then the display driver sample can have additional handling with `display_flush` if that flag is set (same thing for LVGL).

How would `screen_info` help here? Can't the UI always issue the `flush` calls, the display can simply ignore them? In what situation would a `screen_info` flag solve a problem we have? I'm sure I'm just not seeing it.

I can try to integrate it into eLCDIF. What would a good dev board be to test it?

Finomnis commented:
> If you have access to an RT1040, RT1050, RT1060, or RT1064 EVK those are probably best. If not, I can block some time out to put together a basic implementation

Those don't come with a display though, right?

Finomnis commented:
> If you have access to an RT1040, RT1050, RT1060, or RT1064 EVK those are probably best. If not, I can block some time out to put together a basic implementation

Ii might have a 1064 EVK at work, but not the display.

danieldegrasse commented:
> Ii might have a 1064 EVK at work, but not the display.

Hmm, ok. If you can make the API changes then I can work on putting together a change for the LCDIF driver to use this API.

@pdgendt I'd like your thoughts (since I know you've used the RT parts with the LCDIF in the past)- does the change as described here: https://github.com/zephyrproject-rtos/zephyr/pull/79936#issuecomment-2427091733 make sense to you?

Finomnis commented:
@danieldegrasse The reason I'm hesitant is because I don't see what the `screen_info` flag would solve.

Even if the UI manager would know that the display supports flushing, there still wouldn't be a switch case in the UI. I would implement it in a way that it would simply always issue flush, no matter what the display capabilities are, because in the worst case, the flush simply doesn't do anything. So the flag doesn't really carry any useful information to the UI.

Further, lets say it's description is "when this flag is set, the display expects a `flush` call before it shows its content on screen".
That immediately breaks all existing UI managers, meaning, it won't show anything on screen. (Except of LVGL, which we patch here to behave that way).

The reason this happens is because this flag introduces a "MUST" requirement. Like, "if this flag is set, the UI MUST issue a flush call". Backwards compatibility only works with "MAY" requirements.

Like, if we were to introduce a way for the UI manager to tell the display "I am capable of issuing flush calls, you MAY buffer the pixel data and not display it right away", then everything would be fine, and all existing UI managers simply would not tell the display that it is allowed to do so. But I don't think there's a way for this kind of information to get passed from UI to screen.

One way to do so would be to introduce another `disable_immediate_mode` API function, which tells the display that it is allowed to buffer. The display may simply ignore that and stay in immediate mode, and then ignore subsequent flush calls, so everything is fine (= backwards compatibility of displays). UIs can simply not issue it and displays stay in immediate mode (= backwards compatibility of UIs). But if both are updated to the new API, the UI can issue that function and the display understands it and we have the new sync behaviour.

Finomnis commented:
@danieldegrasse What do you think of this change?

```diff
diff --git a/include/zephyr/drivers/display.h b/include/zephyr/drivers/display.h
index 4ee8e66733c..fb066afb43a 100644
--- a/include/zephyr/drivers/display.h
+++ b/include/zephyr/drivers/display.h
@@ -127,6 +127,8 @@ struct display_buffer_descriptor {
        uint16_t height;
        /** Number of pixels between consecutive rows in the data buffer */
        uint16_t pitch;
+       /** The data may be delayed until a `flush` happens */
+       bool allow_buffering;
 };
 
 /**
@@ -290,8 +292,14 @@ static inline int display_read(const struct device *dev, const uint16_t x,
 /**
  * @brief Indicates that all previous writes should be flushed to the display
  *
- * Can be used to present all previous writes simultaneously, for
- * double buffered or latched displays.
+ * Can be used in conjunction with `display_buffer_descriptor::allow_buffering`
+ * to present multiple writes simultaneously.
+ *
+ * This may allow UIs to render in multiple parts without causing tearlines,
+ * if the display supports it.
+ *
+ * The display may ignore this function and present the given pixels
+ * immediately, regardless of `allow_buffering`.
  *
  * @param dev Pointer to device structure
  *
```

Finomnis commented:
@danieldegrasse How about this?

```diff
diff --git a/include/zephyr/drivers/display.h b/include/zephyr/drivers/display.h
index 4ee8e66733c..892f77d2f6b 100644
--- a/include/zephyr/drivers/display.h
+++ b/include/zephyr/drivers/display.h
@@ -163,6 +163,13 @@ typedef int (*display_read_api)(const struct device *dev, const uint16_t x,
                                const struct display_buffer_descriptor *desc,
                                void *buf);

+/**
+ * @typedef display_allow_buffering_api
+ * @brief Callback API to indicate that the display may enter buffered mode.
+ * See display_allow_buffering() for argument description
+ */
+typedef int (*display_allow_buffering_api)(const struct device *dev);
+
 /**
  * @typedef display_flush_api
  * @brief Callback API to indicate that all previous writes should be flushed to the display
@@ -229,6 +236,7 @@ __subsystem struct display_driver_api {
        display_blanking_off_api blanking_off;
        display_write_api write;
        display_read_api read;
+       display_allow_buffering allow_buffering;
        display_flush_api flush;
        display_get_framebuffer_api get_framebuffer;
        display_set_brightness_api set_brightness;
@@ -287,11 +295,35 @@ static inline int display_read(const struct device *dev, const uint16_t x,
        return api->read(dev, x, y, desc, buf);
 }

+/**
+ * @brief Allows buffering pixel data until flushed.
+ *
+ * Can be used in conjunction with display_flush()
+ * to present multiple writes simultaneously.
+ *
+ * This allows UIs to render in multiple parts without causing tearlines,
+ * if the display supports it.
+ *
+ * The display may choose ignore this function and stay in immediate mode.
+ *
+ * @param dev Pointer to device structure
+ *
+ * @retval 0 on success else negative errno code.
+ */
+static inline int display_allow_buffering(const struct *dev){
+       struct display_driver_api *api = (struct display_driver_api *)dev->api;
+
+       if (api->allow_buffering == NULL) {
+               return -ENOSYS;
+       }
+
+       return api->allow_buffering(dev);
+}
+
 /**
  * @brief Indicates that all previous writes should be flushed to the display
  *
- * Can be used to present all previous writes simultaneously, for
- * double buffered or latched displays.
+ * See display_allow_buffering() for more details.
  *
  * @param dev Pointer to device structure
  *
```

Finomnis commented:
@danieldegrasse What do you think about my previous two messages? Should I still go ahead and implement the `screen_info` flag?

faxe1008 commented:
@Finomnis you are right in that it does require downstream changes. What I meant was that we do not break the `write` method parameters downstream. From my point of view the flag serves to document the requirement on an API level. Sure you can just do the flushing anyways but this way you might have to check for `ENOSYS` explicitly additionally for any error checking. Since the `screen_info` still has plenty of room I'd favor making downstream error checking less of a hassle.

Finomnis commented:
Got it. Will implement the suggested changes.

Finomnis commented:
@danieldegrasse @faxe1008 Is this going in the right direction now?

Finomnis commented:
> Thanks for the explanation @Finomnis. Change itself looks good. Can you maybe add a short sentence about this to the v4.0 release notes?

Added release notes.

JarmouniA reviewed (commented):
"present" is a little bit confusing (can be understood as "display is existent/available"), I think "show" would be a better & less confusing term at first glance.

Finomnis commented:
> "present" is a little bit confusing (can be understood as "display is existent/available"), I think "show" would be a better & less confusing term at first glance.

I see :) Any contrary opinions?

Finomnis commented:
> "present" is a little bit confusing (can be understood as "display is existent/available"), I think "show" would be a better & less confusing term at first glance.

@JarmouniA Then `SCREEN_INFO_REQUIRES_SHOW`? Or do you have a better name for this flag?

pdgendt commented:
How about "render"?

danieldegrasse commented on doc/releases/release-notes-4.0.rst:
This is ok for now, but this feature will have to go into 4.1 (once 4.0 releases) as we are now in the 4.0 feature freeze.

Finomnis replied:
Will there be a 4.1 release notes soon? Then I'll add it there

dkalowsk replied:
@Finomnis we're working on getting those merged in

Finomnis replied:
Rebased on main and moved the release notes into 4.1.

danieldegrasse commented on include/zephyr/drivers/display.h:
```diff
@@ -89,6 +89,10 @@ enum display_screen_info {
 	 */
 	SCREEN_INFO_X_ALIGNMENT_WIDTH	= BIT(4),
+	/**
+	 * Screen will only present updated pixels after being told so
```
Maybe expand this here, to something like so:

```
Screen will not show updated framebuffer to user until @ref display_show is called
```

Finomnis replied:
Fixed

danieldegrasse reviewed (commented):
Looking good so far- I think we also want to add usage of this flag into the display sample `samples/drivers/display`

faxe1008 reviewed (commented):
Apart from the small docstring mentioned by @danieldegrasse , looking very good 👍

Finomnis committed:
samples: drivers: display: Add `display_show`

Displays with the capability `SCREEN_INFO_REQUIRES_SHOW`
now require `display_show` to be called before the new frame
is shown on screen.

Signed-off-by: Martin Stumpf <finomnis@gmail.com>

Finomnis commented:
> Looking good so far- I think we also want to add usage of this flag into the display sample `samples/drivers/display`

Added `display_show` to display driver sample.

faxe1008 commented:
Check the compliance issues
https://github.com/zephyrproject-rtos/zephyr/actions/runs/11583276499/job/32248116967#step:11:46
Otherwise LGTM

Finomnis commented:
> Check the compliance issues https://github.com/zephyrproject-rtos/zephyr/actions/runs/11583276499/job/32248116967#step:11:46 Otherwise LGTM

Fixed

Finomnis commented:
@pdgendt You still have a "change requested" open that has already been done :)

JarmouniA reviewed (approved):
Great work!

pdgendt reviewed (dismissed):
> @pdgendt You still have a "change requested" open that has already been done :)

Sorry, yeah this LGTM!

Finomnis commented:
Still blocked by @danieldegrasse's request to add an implementation of this to the eLCDIF peripheral.

Finomnis commented:
@danieldegrasse yes, the idea is indeed that the buffer is reusable after the write. Otherwise you could not do multiple writes before issuing a `show`.

danieldegrasse reviewed (dismissed):
I am slightly concerned that we don't have a full implementation here, but I don't want to block over that (and I'm not aware of any IP upstream that fits this model)

Given that the eLCDIF wouldn't really fit here, I'm going to go ahead and approve- this does need to wait until after 4.0 to merge though

Finomnis commented:
@danieldegrasse One example would be the EA DOGM132W-5 display that contains an ST7565R display driver chip. It is a 132*32 display, but the display internally can hold 132*64 pixels. And you can tell the controller which row in memory the first line of the image is. So with the new `show` api you could flip between the two memory regions and render the entire display with a frame buffer of 132 bytes, and no tearing.

Finomnis commented:
@danieldegrasse We could implement it in `zephyr,sdl-dc`. That one uses the exact structure we require, it does:
```c
	SDL_UpdateTexture(texture, &rect, buf, 4 * rect.w);

	if (display_on) {
		SDL_RenderClear(renderer);
		SDL_RenderCopy(renderer, texture, NULL, NULL);
		SDL_RenderPresent(renderer);
	}
```

The entire `if (display_on) { ... }` should get moved into a function that only gets called when `display_show` is called.

Finomnis committed:
drivers: display: display_sdl: implement display_show

Adds frame synchronization to every frame.
This prevents frame tearing.

Signed-off-by: Martin Stumpf <finomnis@gmail.com>

Finomnis commented:
@danieldegrasse Implemented it in `drivers/display/display_sdl`.

To see the difference, I added a `k_msleep(10)` to every `display_write` call.

Before:
![without_view](https://github.com/user-attachments/assets/f3addb3c-7f14-42fc-b934-f8a09a23371b)

After:
![with_view](https://github.com/user-attachments/assets/1d40fed4-f619-4d58-afa0-ee2ba9b06616)

You can see, the total framerate is the same, bit after the change, the frames pop into existance, while before, you can see the "rolling shutter" effect of the partial writes.

faxe1008 commented:
@Finomnis good idea to use the sdl display for it :+1:

Finomnis commented on doc/releases/release-notes-4.1.rst:
```diff
@@ -94,6 +94,7 @@ Drivers and Sensors
 
   * Added API function :c:func:`display_show` and display capability :c:enum:`SCREEN_INFO_REQUIRES_SHOW`
     for displays that allow tearing prevention of segmented updates. (:github:`79936`)
+  * Added :c:func:`display_show` capability SDL display driver (:dtcompatible:`zephyr,sdl-dc`) make it :c:enum:`SCREEN_INFO_REQUIRES_SHOW`. (:github:`79936`)
```
Adjusted; the entire sentence was wanky imo

pdgendt commented on doc/releases/release-notes-4.1.rst:
```diff
@@ -92,6 +92,10 @@ Drivers and Sensors
 
 * Display
 
+  * Added API function :c:func:`display_show` and display capability :c:enum:`SCREEN_INFO_REQUIRES_SHOW`
+    for displays that allow tearing prevention of segmented updates. (:github:`79936`)
+  * Implemented :c:func:`display_show` for SDL display driver (:dtcompatible:`zephyr,sdl-dc`) and added display capability :c:enum:`SCREEN_INFO_REQUIRES_SHOW`. (:github:`79936`)
```
Nit: try to keep the line length <= 100 chars.

Finomnis replied:
Fixed

danieldegrasse commented:
> @Finomnis good idea to use the sdl display for it 👍

Agreed, thank you for doing this!

Something I was considered regarding the API as we have it now- what `display_show` effectively indicates to the display driver is that the rendering framework (LVGL or otherwise) is done with the current frame.

Another use case I'm looking at right now that would benefit from this information is the "tearing enable" signal used by some displays. The idea of this signal is that it indicates when the display controller will start reading from its internal graphics ram- so if used correctly, the MCU can time writes to that graphics RAM so that the read pointer of the display controller never overlaps with the write pointer from the MCU, and the tearing effect is avoided.

Here's a timing diagram that makes this a bit more clear (from the ST7796S controller datasheet):
![image](https://github.com/user-attachments/assets/e5557b27-5f21-4baf-af7e-06e161bbcc3b)

When using a framebuffer that only covers part of the screen (IE 1/10), we would ideally like to stream all the framebuffers for a given frame after one TE output signal- to do this, the display controller needs a way to know which framebuffer is the last one in the given frame.

We could use `display_show` for this, but we would be adding this API to nearly every display in Zephyr- the TE signal is very common across SPI displays. 

I know we previously discussed extending the `display_write` API in some way- have we considered extending the `display_buffer_descriptor` structure to add a `last` field? This wouldn't break downstream applications, and would allow LVGL or other frameworks to signal to display drivers which frame is the last.

Finomnis commented:
@danieldegrasse I specifically don't want to add it to `write`, because it's asynchronous to that.

Meaning, you can reuse the partial frame buffer already while `show` is still processing. This is especially useful in LVGL threaded present mode, where the render thread can continue while the 'present' thread still runs `show`.

That said, this is already possible with the current api, no? Just add a boolean that tracks if we are the first `write` after a `show`.

On the other hand, if you say "let's not introduce an API that breaks downstream GUI frameworks", that would annoy me a little, because that's exactly what I was saying a couple of weeks ago, and where other people here said it's fine.

Finomnis commented:
> @danieldegrasse The reason I'm hesitant is because I don't see what the `screen_info` flag would solve.
> 
> Even if the UI manager would know that the display supports flushing, there still wouldn't be a conditional based on it in the UI. I would implement it in a way that it would simply always issue flushes, no matter what the display capabilities are, because in the worst case, the flush simply doesn't do anything. So the flag doesn't really carry any useful information to the UI.
> 
> Further, lets say its description is "when this flag is set, the display expects a `flush` call before it shows its content on screen". That immediately breaks all existing UIs, meaning, it won't show anything on screen. (Except of LVGL, which we patch here to behave that way).
> 
> The reason this happens is because this flag introduces a "MUST" requirement. Like, "if this flag is set, the UI MUST issue a flush call". Backwards compatibility only works with "MAY" requirements.
> 
> Like, if we were to introduce a way for the UI manager to tell the display "I am capable of issuing flush calls, you MAY buffer the pixel data and not display it right away", then everything would be fine, and all existing UI managers simply would not tell the display that it is allowed to do so.
> 
> One way to do so would be to introduce another `disable_immediate_mode` or `allow_buffering` API function, which tells the display that it is allowed to buffer. The display may simply ignore that and stay in immediate mode, and then ignore subsequent flush calls, so everything is fine (= backwards compatibility of displays). On the other hand, UIs can simply not issue it and displays stay in immediate mode (= backwards compatibility of UIs). But if both are updated to the new API, the UI can issue that function and the display understands it and we have the new sync behaviour.
> 
> The disadvantage of such an approach would be that it would change the internal state of the display driver permanently. Maybe a better alternative would be to add an `allow_buffering` to the `display_buffer_descriptor` that is disabled by default. That would solve the corner case where multiple UIs write to a display (maybe a split framebuffer with two UIs, like PiP mode?) and one UI has the capability and the other doesn't. **EDIT:** Actually, I retract the suggestion of adding it to the `display_buffer_descriptor`. It introduces undefined behavior in existing UIs that do not initialize that field. I think that adding a second API function that opts-in to this behavior would be the better solution.

@danieldegrasse That's kind of exactly what I was talking about here, and what was dismissed.

Finomnis commented:
> Yeah, this is a great explanation- thank you for laying it out. It seems like LVGL has `lv_display_flush_is_last` for this use case. Personally I see the reasoning for this API- I'm not sure how else we can do this effectively with partial rendering. Maybe we could add a flag to `display_write`? I think a new API is better here though, since we don't break downstream applications that way.

@danieldegrasse Also you said this.

Finomnis commented:
@danieldegrasse Summary:

- I'm fine with either
- Advantage of `display_buffer_descriptor` flag is that it doesn't break anything (assuming uninitialized fields in structs get zeroed automatically)
- Advantage of current approach is that `display_show` can get called while the previous `write` buffer gets reused elsewhere already

Disregarding the asynchronism advantage, I think adding a flag in `display_buffer_descriptor` has many advantages. Although a combination of both can achieve both advantages without disadvantages; with the down side of being a little complicated for the api user.

danieldegrasse commented:
> @danieldegrasse I specifically don't want to add it to write, because it's asynchronous to that.
> 
> Meaning, you can reuse the partial frame buffer already while show is still processing. This is especially useful in LVGL threaded present mode, where the render thread can continue while the 'present' thread still runs show.

Okay, I follow- essentially `display_show` may take some time, but it should not block LVGL from using rendering buffers. The reasoning makes sense here.

> @danieldegrasse That's kind of exactly what I was talking about here, and what was dismissed.
> 
> So I though the fact that displays will no longer be backward compatible with existing GUI managers is something everyone understands and accepts.

I apologize if I or anyone else came off as dismissive here- I think the concern is that we wanted to avoid a hard migration- that is changing the `display_write` API outright. With a flag like the PR uses now, GUI managers will work fine with existing displays, and will only need to be updated gradually as more displays use this API.


> My previous suggestion was to add a allow_buffering field or is_not_last field to the display_buffer_descriptor, so that by default old GUIs simply issue a present in every write.
> 
> That has the drawback, though, that every GUI that doesn't explicitly zero this bit technically leaves it uninitialized, and hence causes undefined behavior. I think it will get zeroed in most cases, but it still is undefined behavior in C I think.

I think this is undefined behavior, yeah. I agree though that modifying the `display_buffer_descriptor` seems like an attractive option.

To be clear, I'm not opposed to the current state of the PR- I just wanted to engage because I realized that displays using the tearing enable signal will also benefit from the GUI framework communicating the "last frame" to them in some way- the reason I leaned towards the display buffer descriptor is that the MIPI DBI layer is used for SPI based displays- if we extended the display buffer descriptor, we could pass the "last frame" flag to the MIPI DBI controller (where it would actually be used) very easily

Finomnis commented:
> Okay, I follow- essentially `display_show` may take some time, but it should not block LVGL from using rendering buffers. The reasoning makes sense here.

@danieldegrasse To be honest, the more I think about it the less this seems like an actual advantage. You can already use double buffering and the additional present thread for that; so there is already some kind of asynchronism.

I'm kind of leaning towards a refactor. Although at this point I might even create a new PR as it will be completely different.

danieldegrasse commented:
> Disregarding the asynchronism advantage, I think adding a flag in display_buffer_descriptor has many advantages. Although I would negate it so that the default 'false' value does not break new display implementations, like has_continuation or similar.

>  So your decision: Should I refactor? I can do it, if requested, no problem. That will also make this change a lot smaller.

I personally am in favor of using a flag in the `display_buffer_descriptor`, but I want to avoid unnecessary work for you, so let's get a few other opinions.

To summarize- we have the following use cases:
- displays that write data to an external framebuffer that will only be displayed once requested. These displays need to know when the entire framebuffer is sent so they can explicitly display it.
- displays that write data to an external framebuffer that is displayed on a fixed interval. These displays need to know when the entire framebuffer is sent so they can time the next write to be after the current framebuffer has been displayed.

The API changes we'd add to enable this:
- add a "bufferable" flag to the `struct display_buffer_descriptor`. When set, this indicates the display buffer can be buffered, and is not the final buffer in the frame. If clear, the display controller should write this buffer directly to the display.

GUI frameworks can use this feature by setting the "bufferable" flag for all writes except the final write in a display frame. This way, the controller is effectively aware which write is the final one, and can take appropriate action.

The only open issue here to solve is how to deal with GUI frameworks that have not been updated and do not initialize the "bufferable" flag- the flag would be an undefined value if the display buffer descriptor was declared on the stack.

@Finomnis does this seem like a good summary?

Finomnis commented:
> @Finomnis does this seem like a good summary?

@danieldegrasse Yes, I think so.

> The only open issue here to solve is how to deal with GUI frameworks that have not been updated and do not initialize the "bufferable" flag- the flag would be an undefined value if the display buffer descriptor was declared on the stack.

In the case they do

```c
struct display_buffer_descriptor desc = {
    .member = foo,
    .member2 = foo2,
};
```

then it should work, I think, I think missing members do get zeroed. (Which is how we use most of our APIs; if drivers don't implement an API function we assume it is a null pointer)

However, if they do

```c
struct display_buffer_descriptor desc;
desc.member = foo;
desc.member2 = foo2;
```

... then I think it's undefined behavior. Which is why I strongly favor the first version everywhere I can.

danieldegrasse commented:
> The only open issue here to solve is how to deal with GUI frameworks that have not been updated and do not initialize the "bufferable" flag- the flag would be an undefined value if the display buffer descriptor was declared on the stack.

One option here is to add a check to all display drivers that throws an error if the "bufferable" flag is set- as display drivers enable support for this, we would remove this check. Not sure how I feel about that- it would quickly "catch" GUI managers that were not initializing the flag, but once we add support for the "bufferable" flag in a display driver then it would be impossible to tell the difference between a GUI framework that recognized the flag and was setting it, and one that was leaving it uninitialized.

danieldegrasse commented:
> @danieldegrasse Yes, I think so.

Great. Again, I apologize for the churn here- I should have considered the tearing enable case a bit more during my initial review. @pdgendt, @faxe1008, @JarmouniA  and @jfischer-no do you (or anyone else) have thoughts on this change?

danieldegrasse commented:
> ... then I think it's undefined behavior. Which is why I strongly favor the first version everywhere I can

Yeah, I believe it is. I'll note that structures have been extended like this in the past- see an example here #77502

Finomnis commented:
@danieldegrasse be aware that you want to wait before the first frame of the image, not after the last, in your TE case. Otherwise the GUI needs to render the frame after the TE and the data will come later than desired.

danieldegrasse commented:
> @danieldegrasse be aware that you want to wait before the first frame of the image, not after the last, in your TE case. Otherwise the GUI needs to render the frame after the TE and the data will come later than desired.

Sure, thanks for the clarification- I'm aware that display drivers would still need some kind of flag to indicate "the last framebuffer we received wasn't bufferable, so this next framebuffer should be sent at the TE interval"

Finomnis commented:
> Sure, thanks for the clarification- I'm aware that display drivers would still need some kind of flag to indicate "the last framebuffer we received wasn't bufferable, so this next framebuffer should be sent at the TE interval"

This is very confusing to read :D maybe we should not call it `bufferable`.

Maybe `frame_incomplete`, `incomplete`, `not_last`, something like that? What's your opinion, @danieldegrasse?

danieldegrasse commented:
> Maybe frame_incomplete, incomplete, not_last, something like that? What's your opinion, @danieldegrasse?

Sure, I like `frame_incomplete`. I'll edit the initial comment with that name

Finomnis commented:
@danieldegrasse The only thing I dislike about this solution is that `display_buffer_descriptor` is also used for `read`, for which it makes no sense. So it feels like the wrong place, although it's the most elegant solution so far I think.

Finomnis commented:
@danieldegrasse For the rework, could you take a look at `lvgl_display_mono.c` and explain it to me? Specifically:
- how `display_blanking_on` and `display_blanking_off` is dealt with here (misuse if blanking? imo this should be removed, as drivers can now easily use `frame_incomplete` to achieve the same internally)
- `SCREEN_DOUBLE_BUFFER`? It seems like that one is just completely random, it just writes the exact same data twice; why?? That's now what I know as 'double buffering' ... or am I missing something?

Finomnis commented:
@danieldegrasse Rework at #81250.

danieldegrasse commented:
> ... just that they didn't realize that this is something more generic, and instead hard-coded it to the specific usecase ...
>
> Should I keep it or remove it?

I'm less familiar with e-ink displays, but I would guess that tearing is much more obvious there due to the slower refresh rate. It does seem like the change was intended to prevent flicker: https://github.com/zephyrproject-rtos/lvgl/commit/0db5fca20cfe183063e6a10027ace94624ce7e23

> SCREEN_DOUBLE_BUFFER? It seems like that one is just completely random, it just writes the exact same data twice; why?? That's not what I know as 'double buffering' ... or am I missing something? If that is meant to write to a back buffer, then flip and write to a front buffer (to keep two buffers synced), then this will definitely break with the introduction of frame_incomplete ...

I'll be honest, this one stumps me. It has been present in the LVGL port since the initial port to Zephyr: https://github.com/zephyrproject-rtos/zephyr/commit/cbfcae7a1f40bdf37ed18a16e23c563a07da9e99#diff-aff42d60cfd55aebd4f8e00f4415a92a6a40d300a0b673cd381d4a3ba6ed4377R28

The only driver that *was* using it was the SSD1673: https://github.com/zephyrproject-rtos/zephyr/commit/49ffca45539041f4307ea4b931f51f70bcce2149#diff-9d44933c67e307bb75712984b2aa52ed1b07a4feb1edd6bd1674c0fe6cc6c1b0R339.

No drivers in tree currently use it:
```
rg SCREEN_INFO_DOUBLE_BUFFER .

./modules/lvgl/lvgl_display_mono.c
37:     if (data->cap.screen_info & SCREEN_INFO_DOUBLE_BUFFER) {

./include/zephyr/drivers/display.h
87:     SCREEN_INFO_DOUBLE_BUFFER       = BIT(3),
```

I *suspect* that the double buffer flag might have been related to this code that exists in the current version of the SSD1673 driver: https://github.com/zephyrproject-rtos/zephyr/blob/c50777a8431157bd4ab3a2c2e2a37e8151ae07da/drivers/display/ssd16xx.c#L484 It seems that the driver will swap between two buffers when using a partial refresh, so two writes are needed to be sure the same framebuffer is in use for both the front and back buffer.


IMO, both of these should be removed (thank you for catching them), but we don't have to do it as part of this PR- we can log an issue against these and remove them in the future

Finomnis commented:
Superseded by #81250.
